cpp/src/shuffler/shuffler.cpp

```diff
@@ -170,195 +172,227 @@ class Shuffler::Progress {
-            );
-            if (chunk.concat_data_size() > 0) {
+        {
+            auto const t0_send_metadata = Clock::now();
+            auto ready_chunks = shuffler_.outgoing_postbox_.extract_all_ready();
+            RAPIDSMPF_NVTX_SCOPED_RANGE("meta_send", ready_chunks.size());
+            for (auto&& chunk : ready_chunks) {
+                // All messages in the chunk maps to the same key (checked by the PostBox)
+                // thus we can use the partition ID of the first message in the chunk to
+                // determine the source rank of all of them.
+                auto dst = shuffler_.partition_owner(shuffler_.comm_, chunk.part_id(0));
+                log.trace("send metadata to ", dst, ": ", chunk);
                 RAPIDSMPF_EXPECTS(
-                    outgoing_chunks_.insert({chunk.chunk_id(), std::move(chunk)}).second,
-                    "outgoing chunk already exist"
+                    dst != shuffler_.comm_->rank(), "sending chunk to ourselves"
                 );
+
+                fire_and_forget_.push_back(shuffler_.comm_->send(
+                    chunk.serialize(), dst, metadata_tag, shuffler_.br_
+                ));
+                if (chunk.concat_data_size() > 0) {
+                    RAPIDSMPF_EXPECTS(
+                        outgoing_chunks_.insert({chunk.chunk_id(), std::move(chunk)})
+                            .second,
+                        "outgoing chunk already exist"
+                    );
+                }
             }
+            stats.add_duration_stat(
+                "event-loop-metadata-send", Clock::now() - t0_send_metadata
+            );
         }
-        stats.add_duration_stat(
-            "event-loop-metadata-send", Clock::now() - t0_send_metadata
-        );
 
         // Receive any incoming metadata of remote chunks and place them in
         // `incoming_chunks_`.
-        auto const t0_metadata_recv = Clock::now();
-        while (true) {
-            auto const [msg, src] = shuffler_.comm_->recv_any(metadata_tag);
-            if (msg) {
-                auto chunk = Chunk::deserialize(*msg, false);
-                log.trace("recv_any from ", src, ": ", chunk);
-                // All messages in the chunk maps to the same Rank (checked by the
-                // PostBox) thus we can use the partition ID of the first message in the
-                // chunk to determine the source rank of all of them.
-                RAPIDSMPF_EXPECTS(
-                    shuffler_.partition_owner(shuffler_.comm_, chunk.part_id(0))
-                        == shuffler_.comm_->rank(),
-                    "receiving chunk not owned by us"
-                );
-                incoming_chunks_.insert({src, std::move(chunk)});
-            } else {
-                break;
+        {
+            auto const t0_metadata_recv = Clock::now();
+            RAPIDSMPF_NVTX_SCOPED_RANGE("meta_recv", -1);
```

Comment:
The `-1` value is just to fill it because it's a required argument, right? Assuming there's a proper NVTX API for that, maybe a better solution is adding an alternative scoped range implementation that doesn't take a value instead, having a value here implies the user needs to read the code to figure out what that means.